opendevin/controller.py

```diff
@@ -0,0 +1,53 @@
+from opendevin.lib.command_manager import CommandManager
+from opendevin.lib.event import Event
+
+MAX_OUTPUT_LENGTH = 5000
+
+def print_callback(event):
+    print(event, flush=True)
+
+class AgentController:
+    def __init__(self, agent, max_iterations=100, callbacks=[]):
+        self.agent = agent
+        self.max_iterations = max_iterations
+        self.background_commands = []
+        self.command_manager = CommandManager()
+        self.callbacks = callbacks
+        self.callbacks.append(self.agent.add_event)
+        self.callbacks.append(print_callback)
+
+    def maybe_perform_action(self, event):
+        if not (event and event.is_runnable()):
+            return
+        action = 'output'
+        try:
+            output = event.run(self)
+        except Exception as e:
+            output = 'Error: ' + str(e)
+            action = 'error'
+        if len(output) > MAX_OUTPUT_LENGTH:
+            output = output[:MAX_OUTPUT_LENGTH] + '...'
+        out_event = Event(action, {'output': output})
+        return out_event
```

Comment:
Do you mean the truncation logic specifically? I agree it'd be better to leave that to the agent--I will factor that part out.

But for the actions themselves, yes I think it's better to give the agent a "menu" of possible actions, and leave action execution outside of the agent. For two reasons:
* it simplifies the agent's contract--the agent just has to translate natural language into a stream of Events
* it makes things predictable--the UI can reflect _exactly_ what the agent is doing at any given time. IMO this part is super important: the user is going to want a high degree of visibility into how the AI is behaving. E.g. if there's an `error` event, we show it in red, if there's a `browse` event we bring the browser in focus, etc.

Reply:
For UI purposes, how about we expect each agent to generate an `Event` (when each action that need to be tracked by UI is triggered - as discussed above) but do not assume how these `Event` will be generated by the `Agent`?

I worry about the fixed action space is because CodeAct agent's action space is just bash commands - if we make it fixed, we might have trouble support different agent variants.